fluent/client/ws/connection_test.go

```diff
@@ -0,0 +1,194 @@
+package ws_test
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/IBM/fluent-forward-go/fluent/client/ws"
+	"github.com/IBM/fluent-forward-go/fluent/client/ws/ext/extfakes"
+	"github.com/gorilla/websocket"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+type message struct {
+	mt  int
+	msg []byte
+	err error
+}
+
+var _ = Describe("Connection", func() {
+	var (
+		fakeConn           *extfakes.FakeConn
+		opts               ws.ConnectionOptions
+		readMsgs           chan message
+		rhCallCt, rmCallCt int
+		onFail             func(err error)
+	)
+
+	BeforeEach(func() {
+		rmCallCt = 0
+		rhCallCt = 0
+		readMsgs = make(chan message, 1)
+		fakeConn = &extfakes.FakeConn{}
+		fakeConn.ReadMessageStub = func() (int, []byte, error) {
+			rmCallCt++
```

Comment:
I'm not sure how to avoid some of these races. I don't think they're all valid. I added some changes to this and another case, but otherwise, I think the ROI is low.

Reply:
Ideally we could run `go test -race` to detect real races (if these currently are only races in the test file) and not exit on first encountered race.

I suppose for now, we probably need to open an issue and add the build tag to ignore races from this file. 
See https://golang.org/doc/articles/race_detector#Excluding_Tests

Adding this tag to the ws_test.go will ignore all races from failing our test. This will work on the entire file, unfortunately, not at the specific test-case level
`// +build !race`

Reply:
All the remaining races that I see are within the tests themselves. It's a lot of noise but might be useful if things get flaky.